File: scripts/scrape_roadmap_bed_files.py
#!/usr/bin/env python3
"""
Web scraping script to extract download links for all .bed files from
Roadmap Epigenomics DNase promoter BED files directory.

URL: https://egg2.wustl.edu/roadmap/data/byDataType/dnase/BED_files_prom/
"""

import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlparse
import sys
from typing import List, Set


def get_bed_file_links(url: str) -> List[str]:
    """
    Scrape the webpage and extract all .bed file download links.
    
    Args:
        url: The URL of the directory listing page
        
    Returns:
        List of full URLs to .bed files (excluding .bed.gz files)
    """
    try:
        # Send GET request to the URL
        print(f"Fetching page: {url}")
        response = requests.get(url, timeout=30)
        response.raise_for_status()  # Raise an exception for bad status codes
        
        # Parse the HTML content
        soup = BeautifulSoup(response.content, 'html.parser')
        
        # Find all anchor tags (links)
        links = soup.find_all('a')
        
        bed_urls = []
        base_url = url.rstrip('/')
        
        for link in links:
            href = link.get('href', '')
            if not href:
                continue
            
            # Decode URL-encoded characters (e.g., %5F -> _)
            from urllib.parse import unquote
            decoded_href = unquote(href)
            
            # Check if it's a .bed file (but not .bed.gz)
            if decoded_href.endswith('.bed') and not decoded_href.endswith('.bed.gz'):
                # Construct full URL
                if href.startswith('http'):
                    full_url = href
                else:
                    # Handle relative URLs
                    full_url = urljoin(base_url + '/', href)
                
                bed_urls.append(full_url)
        
        # Remove duplicates while preserving order
        seen: Set[str] = set()
        unique_bed_urls = []
        for url_item in bed_urls:
            if url_item not in seen:
                seen.add(url_item)
                unique_bed_urls.append(url_item)
        
        return sorted(unique_bed_urls)
    
    except requests.exceptions.RequestException as e:
        print(f"Error fetching the webpage: {e}", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error parsing the webpage: {e}", file=sys.stderr)
        sys.exit(1)


def main():
    """Main function to run the scraper."""
    url = "https://egg2.wustl.edu/roadmap/data/byDataType/dnase/BED_files_prom/"
    
    print("=" * 70)
    print("Roadmap Epigenomics DNase Promoter BED Files Scraper")
    print("=" * 70)
    print()
    
    # Get all .bed file links
    bed_links = get_bed_file_links(url)
    
    if not bed_links:
        print("No .bed files found on the page.")
        sys.exit(1)
    
    print(f"Found {len(bed_links)} .bed file(s):\n")
    
    # Print all links
    for i, link in enumerate(bed_links, 1):
        print(f"{i:3d}. {link}")
    
    print()
    print("=" * 70)
    print(f"Total: {len(bed_links)} .bed file(s)")
    print("=" * 70)
    
    # Optionally save to a file
    output_file = "roadmap_bed_file_links.txt"
    try:
        with open(output_file, 'w') as f:
            for link in bed_links:
                f.write(f"{link}\n")
        print(f"\nLinks saved to: {output_file}")
    except Exception as e:
        print(f"\nWarning: Could not save to file: {e}", file=sys.stderr)
    
    return bed_links


if __name__ == "__main__":
    main()

File: scripts/scrape_roadmap_bed_files.R
#!/usr/bin/env Rscript
# Web scraping script to extract download links for all .bed.gz files from
# Roadmap Epigenomics DNase promoter BED files directory.
#
# URL: https://egg2.wustl.edu/roadmap/data/byDataType/dnase/BED_files_prom/
#
# Usage:
#   Rscript scrape_roadmap_bed_files.R
#   or
#   source("scrape_roadmap_bed_files.R")
#   bed_links <- scrape_roadmap_bed_files()

suppressPackageStartupMessages({
    if (!require("rvest", quietly = TRUE)) {
        stop("Package 'rvest' is required. Install it with: install.packages('rvest')")
    }
    if (!require("httr", quietly = TRUE)) {
        stop("Package 'httr' is required. Install it with: install.packages('httr')")
    }
})

#' Scrape the webpage and extract all .bed.gz file download links
#'
#' @param url The URL of the directory listing page
#' @return Character vector of full URLs to .bed.gz files (excluding plain .bed files)
get_bed_file_links <- function(url) {
    tryCatch({
        # Send GET request to the URL
        cat("Fetching page:", url, "\n")
        
        # Read the HTML page
        page <- read_html(url)
        
        # Find all anchor tags (links)
        links <- page %>%
            html_nodes("a") %>%
            html_attr("href")
        
        # Remove NA values
        links <- links[!is.na(links)]
        
        # Base URL for constructing full URLs
        base_url <- sub("/$", "", url)
        
        bed_urls <- character(0)
        
        for (href in links) {
            if (nchar(href) == 0) next
            
            # Decode URL-encoded characters (e.g., %5F -> _)
            decoded_href <- URLdecode(href)
            
            # Check if it's a .bed.gz file (excluding plain .bed files)
            if (grepl("\\.bed\\.gz$", decoded_href) && !grepl("hg38lift|stateno|expanded|mnemonics", decoded_href)) {
                # Construct full URL
                if (grepl("^https?://", href)) {
                    full_url <- href
                } else {
                    # Handle relative URLs
                    if (startsWith(href, "/")) {
                        # Absolute path from domain root
                        parsed_url <- httr::parse_url(url)
                        full_url <- paste0(parsed_url$scheme, "://", parsed_url$hostname, href)
                    } else {
                        # Relative path
                        full_url <- paste0(base_url, "/", href)
                    }
                }
                
                bed_urls <- c(bed_urls, full_url)
            }
        }
        
        # Remove duplicates while preserving order
        unique_bed_urls <- unique(bed_urls)
        
        # Sort the URLs
        return(sort(unique_bed_urls))
        
    }, error = function(e) {
        cat("Error:", conditionMessage(e), "\n", file = stderr())
        stop("Failed to scrape the webpage")
    })
}

#' Main function to run the scraper
#'
#' @param url The URL to scrape (default: Roadmap Epigenomics DNase promoter, enhancer, and chromhmm directory)
#' @param output_file Optional output file to save links (default: "roadmap_bed_gz_file_links_prom.txt", "roadmap_bed_gz_file_links_enh.txt", "roadmap_bed_gz_file_links_chromhmm.txt")
#' @param verbose Whether to print progress messages (default: TRUE)
#' @return Character vector of .bed.gz file URLs
scrape_roadmap_bed_files <- function(
    url = c("https://egg2.wustl.edu/roadmap/data/byDataType/dnase/BED_files_prom/",
    "https://egg2.wustl.edu/roadmap/data/byDataType/dnase/BED_files_enh/",
    "https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/"),
    verbose = TRUE
) {
    url <- match.arg(url)

    if (verbose) {
        cat(paste(rep("=", 70), collapse = ""), "\n")
        cat(paste("Roadmap Epigenomics ", url, "BED.GZ Files Scraper\n"))
        cat("All files are based on hg19 assembly.\n")
        cat("If you need to use a different assembly, you need to lift over the files to the\n
                desired assembly using `rtracklayer::liftOver()`.\n")
        cat(paste(rep("=", 70), collapse = ""), "\n")
        cat("\n")
    }
    
    # Get all .bed.gz file links
    bed_links <- get_bed_file_links(url)
    
    if (length(bed_links) == 0) {
        cat("No .bed.gz files found on the page.\n")
        return(character(0))
    }
    
    if (verbose) {
        cat("Found", length(bed_links), ".bed.gz file(s):\n\n")
    }
    
    return(bed_links)
}

data("roadmap_metadata")
url <- c("https://egg2.wustl.edu/roadmap/data/byDataType/dnase/BED_files_prom/",
"https://egg2.wustl.edu/roadmap/data/byDataType/dnase/BED_files_enh/",
"https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/")
verbose <- TRUE
roadmap_bed_gz_links_metadata <- lapply(url, function(.url) {
    urls <- scrape_roadmap_bed_files(.url)
    EID <- gsub(".+(E\\d+).*", "\\1", urls)
    urls <- data.frame(EID = EID, URL = urls, stringsAsFactors = FALSE)
    urls <- merge(urls, roadmap_metadata, by.x = "EID", 
                  by.y = "epigenome_id_eid", all.x = TRUE)
})
names(roadmap_bed_gz_links_metadata) <- c("DHS_promoter", 
"DHS_enhancer", "ChromHMM_15_states_5_coreMarks")
saveRDS(roadmap_bed_gz_links_metadata, "roadmap_bed_gz_links_metadata.rds")

usethis::use_data(roadmap_bed_gz_links_metadata, overwrite = TRUE)

